survex: init at 1.2.44

MatthewCroughan opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Adds [survex](https://survex.com/), a package used by cavers in order to survey and map out caves in the real world. [Usually this is only packaged on Debian](https://packages.debian.org/jessie/survex) due to its age and complexity.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

MatthewCroughan committed:
survex: init at 1.2.44

Co-authored-by: Fabian Affolter <mail@fabian-affolter.ch>
Co-authored-by: Christoph Neidahl <christoph.neidahl@gmail.com>

fabaff commented on pkgs/applications/misc/survex/default.nix:
```diff
@@ -0,0 +1,59 @@
+  pname = "survex";
+  version = "1.2.44";
+
+  nativeBuildInputs = [ docbook5 x11 libGL libGLU docbook2x autoreconfHook pkg-config wxGTK30-gtk3 ffmpeg proj python (perl532.withPackages (perlPackages: with perlPackages; [ LocalePO ] )) ];
```
Some inputs looks like that they should go into `buildInputs`.

MatthewCroughan replied:
Probably most of them. I did not understand the distinction between `nativeBuildInputs` and `buildInputs`, now that I do I will correct this.

MatthewCroughan replied:
How's this look? 42e27ff15da3dd376f19008c18a68ada0df854d3

OPNA2608 commented on pkgs/applications/misc/survex/default.nix:
```diff
@@ -0,0 +1,59 @@
+  '';
+
+  postConfigure = ''
+#    substituteInPlace doc/Makefile --replace "docbook2man" "docbook2man --sgml" # Will be needed once sgmltools-lite is packaged.
```
```suggestion
```

MatthewCroughan replied:
Why would you remove the comment? It explains the state of the doc being compiled. Is it considered bad to have comments in Nixpkgs?

OPNA2608 replied:
I think comments are fine, but comments that refer to unpackaged dependencies and possible problems with them are IMO kind of pointless unless you also plan to package that dependency eventually.

If you want to leave it in, then please fix the indentation at least.

MatthewCroughan replied:
The indentation should be just two spaces, correct?

OPNA2608 replied:
Should be:

```suggestion
    # substituteInPlace doc/Makefile --replace "docbook2man" "docbook2man --sgml" # Will be needed once sgmltools-lite is packaged.
```

OPNA2608 commented on pkgs/applications/misc/survex/default.nix:
```diff
@@ -0,0 +1,59 @@
+{ lib
+, stdenv
```
```suggestion
, stdenv
, fetchgit
```

OPNA2608 commented on pkgs/applications/misc/survex/default.nix:
```diff
@@ -0,0 +1,59 @@
+
+  nativeBuildInputs = [ docbook5 x11 libGL libGLU docbook2x autoreconfHook pkg-config wxGTK30-gtk3 ffmpeg proj python (perl532.withPackages (perlPackages: with perlPackages; [ LocalePO ] )) ];
+
+  src = builtins.fetchGit {
+    url = "git://git.survex.com/survex";
+    rev = "38bc41554efda8c15494ca44e5fc17c8146a077f";
+    ref = "refs/tags/${version}";
+  };
```
> while evaluating the attribute 'src' of the derivation 'survex-1.2.44' at <…>/pkgs/stdenv/generic/make-derivation.nix:201:11:
> access to URI 'git://git.survex.com/survex' is forbidden in restricted mode

```suggestion
  src = fetchgit {
    url = "git://git.survex.com/survex";
    rev = version;
    sha256 = "11gaqmabrf3av665jy3mr0m8hg76fmvnd0g3rghzmyh8d8v6xk34";
  };
```

There's a GitHub repo for this project ([ojwb/survex](https://github.com/ojwb/survex)), do you know if this is official? They have already tagged a version 1.2.45 there.

MatthewCroughan replied:
It is an official mirror, yes.

OPNA2608 replied:
Then can we bump the package to 1.2.45 or does something prevent us from doing so?

OPNA2608 commented on pkgs/applications/misc/survex/default.nix:
```diff
@@ -0,0 +1,59 @@
+, python
+, libGL
+, libGLU
+, x11
```
Please don't use aliases in Nixpkgs derivations.

https://github.com/NixOS/nixpkgs/blob/7e529086a7ce588380178733a42ff2a293ac080c/pkgs/top-level/aliases.nix#L875

```suggestion
, xlibsWrapper
```

OPNA2608 replied:
Actually it seems like xlibsWrapper shouldn't be used in general.

https://github.com/NixOS/nixpkgs/blob/15c093a16789a66befd3acb56a6bc18441b7b2b8/pkgs/top-level/all-packages.nix#L18617-L18618

Could you remove it and manually add whatever X11-relevant libraries it asks for?

balsoft commented on pkgs/applications/misc/survex/default.nix:
```diff
@@ -0,0 +1,59 @@
+
+  nativeBuildInputs = [ docbook5 x11 libGL libGLU docbook2x autoreconfHook pkg-config wxGTK30-gtk3 ffmpeg proj python (perl532.withPackages (perlPackages: with perlPackages; [ LocalePO ] )) ];
+
+  src = builtins.fetchGit {
```
Use `fetchgit` from nixpkgs instead of `builtins.fetchGit`.

OPNA2608 commented:
![Screenshot_macOS_2021-06-08_20:20:39](https://user-images.githubusercontent.com/23431373/121241750-12280c80-c89c-11eb-91e8-93c0ee88f2e0.png)

Changes for building on Darwin:

<details>
  <summary>pkgs/data/fonts/unscii/default.nix</summary>

```patch
diff --git a/pkgs/data/fonts/unscii/default.nix b/pkgs/data/fonts/unscii/default.nix
index 066a4d4d922..2bedf6c4f6f 100644
--- a/pkgs/data/fonts/unscii/default.nix
+++ b/pkgs/data/fonts/unscii/default.nix
@@ -2,6 +2,11 @@
 , fontforge, SDL, SDL_image, mkfontscale
 }:
 
+let
+  perlenv = perl.withPackages (p: with p; [
+    TextCharWidth
+  ]);
+in
 stdenv.mkDerivation rec {
   pname = "unscii";
   version = "1.1";
@@ -12,11 +17,23 @@ stdenv.mkDerivation rec {
   };
 
   nativeBuildInputs =
-    [ (perl.withPackages (p: [ p.TextCharWidth ]))
+    [ perlenv
       bdftopcf fontforge SDL SDL_image
       mkfontscale
     ];
 
+  # Fixes shebang -> wrapper problem on Darwin
+  postPatch = ''
+    for perltool in *.pl; do
+      substituteInPlace Makefile \
+        --replace "./$perltool" "${perlenv}/bin/perl ./$perltool"
+    done
+  '';
+
+  makeFlags = [
+    "CC=${stdenv.cc.targetPrefix}cc"
+  ];
+
   preConfigure = ''
     patchShebangs .
   '';
```

</details>

<details>
  <summary>pkgs/applications/misc/survex/default.nix</summary>

```patch
diff --git a/pkgs/applications/misc/survex/default.nix b/pkgs/applications/misc/survex/default.nix
index 1db4b735eb8..24942b3e4fd 100644
--- a/pkgs/applications/misc/survex/default.nix
+++ b/pkgs/applications/misc/survex/default.nix
@@ -4,6 +4,7 @@
 , autoreconfHook
 , pkg-config
 , wxGTK30-gtk3
+, wxmac
 , ffmpeg
 , proj
 , perl532
@@ -14,15 +15,26 @@
 , xlibsWrapper
 , docbook2x
 , docbook5
+, Carbon
+, Cocoa
 }:
 
+let
+  perlenv = perl532.withPackages (perlPackages: with perlPackages; [ LocalePO ] );
+in
 stdenv.mkDerivation rec {
   pname = "survex";
   version = "1.2.44";
 
-  nativeBuildInputs = [ docbook5 docbook2x autoreconfHook pkg-config ];
+  nativeBuildInputs = [ docbook5 docbook2x autoreconfHook pkg-config perlenv python ];
 
-  buildInputs = [ xlibsWrapper libGL libGLU wxGTK30-gtk3 ffmpeg proj python (perl532.withPackages (perlPackages: with perlPackages; [ LocalePO ] )) ];
+  buildInputs = [
+    libGL libGLU ffmpeg proj
+  ] ++ lib.optionals stdenv.hostPlatform.isDarwin [
+    wxmac Carbon Cocoa
+  ] ++ lib.optionals stdenv.hostPlatform.isLinux [
+    wxGTK30-gtk3 xlibsWrapper
+  ];
 
   src = fetchgit {
     url = "git://git.survex.com/survex";
@@ -30,20 +42,25 @@ stdenv.mkDerivation rec {
     sha256 = "11gaqmabrf3av665jy3mr0m8hg76fmvnd0g3rghzmyh8d8v6xk34";
   };
 
+  enableParallelBuilding = true;
+
   # Docs rely on sgmltools-lite, a package that would be quite complex to
   # provide as it is quite old. So this preConfigure hook effectively disables
   # the doc generation. An example of packaging sgmltools-lite from Gentoo can
   # be found here:
   # https://gitweb.gentoo.org/repo/gentoo.git/tree/app-text/sgmltools-lite/sgmltools-lite-3.0.3-r15.ebuild?id=0b8b716331049599ea3299981e3a9ea6e258c5e0
 
-  preConfigure = ''
+  postPatch = ''
+    patchShebangs .
     echo "" > doc/Makefile.am
-  '';
-
-  postConfigure = ''
 #    substituteInPlace doc/Makefile --replace "docbook2man" "docbook2man --sgml" # Will be needed once sgmltools-lite is packaged.
+    for perltool in './extract-msgs.pl' './gettexttomsg.pl' '$(srcdir)/gdtconvert' '$(srcdir)/gen_img2aven'; do
+      substituteInPlace src/Makefile.am \
+        --replace "$perltool" "${perlenv}/bin/perl $perltool"
+    done
+    substituteInPlace lib/Makefile.am \
+      --replace '$(srcdir)/make-pixel-font' '${perlenv}/bin/perl $(srcdir)/make-pixel-font'
     substituteInPlace lib/make-pixel-font --replace /usr/share/unifont/unifont.hex ${unscii.extra}/share/fonts/misc/unifont.hex
-    patchShebangs .
   '';
 
   meta = with lib; {
```

</details>

<details>
  <summary>pkgs/top-level/all-packages.nix</summary>

```patch
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 86eab6704c4..088e1decffe 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -26738,7 +26738,9 @@ in
     git = gitMinimal;
   };
 
-  survex = callPackage ../applications/misc/survex { };
+  survex = callPackage ../applications/misc/survex {
+    inherit (darwin.apple_sdk.frameworks) Carbon Cocoa;
+  };
 
   sunvox = callPackage ../applications/audio/sunvox { };
 
```

</details>

MatthewCroughan commented:
> ![Screenshot_macOS_2021-06-08_20:20:39](https://user-images.githubusercontent.com/23431373/121241750-12280c80-c89c-11eb-91e8-93c0ee88f2e0.png)
> 
> Changes for building on Darwin:
> pkgs/data/fonts/unscii/default.nix
> pkgs/applications/misc/survex/default.nix
> pkgs/top-level/all-packages.nix

That's so great.

MatthewCroughan commented:
@OPNA2608 I've added your patches as a separate commit and squashed all of the previous commits we have made. I think having a clean patch on supporting the Darwin build is important for historical purposes, and it's something cool I can reference to people and show them how Darwin support was added for this.

MatthewCroughan commented:
Actually, I see that you've modified uniscii, that'll have to be modified in a separate commit right?

MatthewCroughan commented on pkgs/applications/misc/survex/default.nix:
```diff
@@ -14,34 +15,52 @@
-
-  postConfigure = ''
     # substituteInPlace doc/Makefile --replace "docbook2man" "docbook2man --sgml" # Will be needed once sgmltools-lite is packaged.
+    for perltool in './extract-msgs.pl' './gettexttomsg.pl' '$(srcdir)/gdtconvert' '$(srcdir)/gen_img2aven'; do
+      substituteInPlace src/Makefile.am \
+        --replace "$perltool" "${perlenv}/bin/perl $perltool"
+    done
+    substituteInPlace lib/Makefile.am \
+      --replace '$(srcdir)/make-pixel-font' '${perlenv}/bin/perl $(srcdir)/make-pixel-font'
```
@OPNA2608 What exactly does this do differently? My solution seems to work, but this is much more complex. What is the benefit from the added complexity?

OPNA2608 replied:
Oh, I wanted to leave a link to the issue but I forgot about it. Same problem as

> `# Fixes shebang -> wrapper problem on Darwin`

in unscii, relevant issue is #65351 (seems to apply to `perl.withPackages` too).

Those perl scripts don't execute properly on Darwin. The platform doesn't seem to like when a shebang for an executable file points to a shell script (like the wrapper that `perl.withPackages` produces so its perl binary can see the packages), so it tries to run those files with the regular shell instead. It's therefore necessary to manually call them with the perl interpreter so their shebangs don't get tried & ignored.

OPNA2608 replied:
If I remove those 6 lines again:

![Screenshot_macOS_2021-06-09_19:14:27](https://user-images.githubusercontent.com/23431373/121399708-fd5c7f00-c956-11eb-9e49-1e80ad79e44d.png)

OPNA2608 commented:
> Actually, I see that you've modified uniscii, that'll have to be modified in a separate commit right?

Yeah.

MatthewCroughan commented:
> > Actually, I see that you've modified uniscii, that'll have to be modified in a separate commit right?
> 
> Yeah.

What would the commit message be for what you have done to unscii?

OPNA2608 commented:
`unscii: Fixed Darwin build`, because it had the same shebang -> wrapper issue with its `perl.withPackages` that survex had.

7c6f434c commented:
Looks reasonable

OPNA2608 committed:
survex: support Darwin build

OPNA2608 committed:
unscii: fix Darwin build

MatthewCroughan commented:
Looks like this can be merged now. Can anybody take a look? :smile:
